woocommerce/woocommerce-ios
# Refactor OrderDetailsViewController and OrderDetailsViewModel

Fixes #1035 

This is a pretty large PR, that requires extensive testing. My apologies for both the length and the extra testing, but there was no way I could think of to make this PR smaller.

I am opening this PR as a draft, first so that I can finally put it out for review, and second so that we have extra room to discuss the way I have refactored the existing code. I considered two alternatives:
1. Separate the view model by "verticals" (on class that fetches shipment trackings and knows how to populate the cells for those trackings, another class that does the same with products...)
2. Make the view model fetch data (products, shipment trackings, etc), pass all that data to a single class that populates all types of cells. This is what I went with in the end, I am still not 100% convinced it is the best possible approach.

## Changes
<img src="https://user-images.githubusercontent.com/2722505/60695692-3d1dbf00-9f15-11e9-9342-f70b2df20da1.jpg" width="350"/>

We would add an extra piece to the existing VC<>VM pair, that would implement `UITableViewDataSource`. 

So, after the refactoring, the responsibilities would be:
- OrderDetailsViewController just renders data and presents Notices
- OrderDetailsViewModel encapsulates data fetching and coordinates data, passing it to a data source
- OrderDetailsDataSource populates cells 

In terms of lines of code, OrderDetailsViewController goes from 1398 lines down to 468, OrderDetailsViewModel grows from 166 to 445 lines. OrderDetailsDataSource is at 772 lines.

OrderDetailsDataSource is still a bit too large, and it can be split even further, but we could do that in a follow up PR, if we decide to follow this approach.

## Testing
- Checkout the branch, navigate to the Order list
- Navigate in an out of orders, both with and without a network connection. 
- Tap all the things: fulfillment, filters, product details... we need to make sure everything works as it used to
